### PR TITLE
adding pareto features

### PR DESCRIPTION
**Pull request by pbougue**
adds visitor (for tracing) and is_dominated method

### CODE REVIEW

**Comment by TeXitoi** [comment]
Please add unit tests.

**Comment by pbougue** [comment]
Tests done

File: multi_obj_pool.h [hunk 1]
```diff
@@ -47,17 +54,22 @@ struct ParetoFront {
...
 
     bool add(const Obj& obj);
+    bool contains_better_than(const Obj& obj);
     template <class Predicate>
     inline void remove_if (Predicate pred) { pareto_front.remove_if(std::move(pred)); }
     inline size_t size() const { return pareto_front.size(); }
     inline const_iterator begin() const { return pareto_front.begin(); }
     inline const_iterator end() const { return pareto_front.end(); }
 
+    const Visitor& get_visitor() const { return visitor; }
```
Comment (TeXitoi) [inline multi_obj_pool.h]: Why const? and If I want to move it out? It'll be simpler if v is just public.

File: tests/multi_obj_pool_test.cpp [hunk 2]
```diff
@@ -36,29 +36,56 @@ www.navitia.io
...
     BOOST_CHECK(! pool.add(2));
     BOOST_REQUIRE_EQUAL(pool.size(), 1);
     BOOST_CHECK_EQUAL(*pool.begin(), 2);
+    BOOST_CHECK_EQUAL(visitor.nb_inserted, 2);
+    BOOST_CHECK_EQUAL(visitor.nb_dominates, 1);
+    BOOST_CHECK_EQUAL(visitor.nb_dominated, 2);
+
+    BOOST_CHECK(pool.contains_better_than(1));
+    BOOST_CHECK(pool.contains_better_than(2));
+    BOOST_CHECK(!pool.contains_better_than(3));
```
Comment (TeXitoi) [inline tests/multi_obj_pool_test.cpp]: `contains_better_than` can be tested with a more complex object to test the weakly dominated case (worst on one dimension, worst on the others).

File: multi_obj_pool.h (hunk 1 above)
Comment (xlqian) [inline multi_obj_pool.h]: @TeXitoi  why you want to move out a visitor?

Comment (TeXitoi) [inline multi_obj_pool.h]: if my visitor have a non copyable element, and after the calculus, I what to return it.

That's pure speculation, of course.

File: multi_obj_pool.h [hunk 3]
```diff
@@ -95,5 +108,19 @@ bool ParetoFront<Obj, Dominator>::add(const Obj& obj) {
...
+    visitor.at_inserted(obj);
     return true;
 }
+
+/*
+ * Returns true if the given solution is dominated by one object of the pareto front
+ */
+template <typename Obj, typename Dominator, typename Visitor>
+bool ParetoFront<Obj, Dominator, Visitor>::contains_better_than(const Obj& obj) {
+    for (const auto& cur : pareto_front) {
```
Comment (xlqian) [inline multi_obj_pool.h]: you mean `return std::any_of()`?

Comment (pbougue) [inline multi_obj_pool.h]: yes that's what I mean :smiley_cat:

**Pull request merged**